File: java/L100/LeetCode0123.java
package L100;

public class LeetCode0123 {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length < 2) {
            return 0;
        }
        final int N = prices.length;
        int[][][] dp = new int[N][3][2];
        for (int i = 0;i < N;++i) {
            for (int j = 0;j < 3;++j) {
                for (int k = 0;k < 2;++k) {
                    dp[i][j][k] = -1000_000;
                }
            }
        }
        dp[0][0][0] = -prices[0]; // 0: 持有股票
        dp[0][0][1] = 0;
        for (int i = 1;i < N;++i) {
            dp[i][0][0] = Math.max(dp[i - 1][0][0], dp[i - 1][0][1] - prices[i]);
            dp[i][0][1] = dp[i - 1][0][1];
            dp[i][1][0] = Math.max(dp[i - 1][1][0], dp[i - 1][1][1] - prices[i]);
            dp[i][1][1] = Math.max(dp[i - 1][1][1], dp[i - 1][0][0] + prices[i]);
            dp[i][2][0] = Math.max(dp[i - 1][2][0], dp[i - 1][2][1] - prices[i]);
            dp[i][2][1] = Math.max(dp[i - 1][2][1], dp[i - 1][1][0] + prices[i]);
        }
        int max = 0;
        for (int i = 0;i < 3;++i) {
            for (int j = 0;j < 2;++j) {
                max = Math.max(max, dp[N - 1][i][j]);
            }
        }
        return max;
    }
}

File: cpp/L100/LeetCode0123.cc
#include <algorithm>
#include <array>
#include <vector>
using std::vector;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() < 2) {
            return 0;
        }
        std::array<std::array<int, 2>, 3> dp{{{0 ,-1000000}, {0 ,-100000000}, {0, -100000000}}};
        for (int price : prices) {
            int t01 = std::max(dp[0][1], dp[0][0] - price);
            int t10 = std::max(dp[1][0], price + dp[0][1]);
            int t11 = std::max(dp[1][1], dp[1][0] - price);
            int t20 = std::max(dp[2][0], price + dp[1][1]);
            int t21 = std::max(dp[2][1], dp[2][0] - price);
            dp[0][1] = t01;
            dp[1][0] = t10;
            dp[1][1] = t11;
            dp[2][0] = t20;
            dp[2][1] = t21;
        }
        int profit = 0;
        for (std::array<int, 2> const &at : dp) {
            for (int x : at) {
                profit = std::max(x, profit);
            }
        }
        return profit;
    }
};